python/cudf/cudf/core/frame.py

```diff
@@ -97,6 +97,133 @@ def _as_column(self):
 
         return self._data[None].copy(deep=False)
 
+    def _scatter(self, key, value):
+        result = self._from_table(libcudfxx.copying.scatter(value, key, self))
+
+        result._copy_categories(self)
+        return result
+
+    def _empty_like(self):
+        result = self._from_table(libcudfxx.copying.table_empty_like(self))
+
+        result._copy_categories(self)
+        return result
+
+    def _slice(self, arg):
+        """
+       _slice : slice the frame as per the arg
+
+       Parameters
+       ----------
+       arg : should always be of type slice and doesn't handle step
+
+       """
+        num_rows = len(self)
+        if num_rows == 0:
+            return self
+        start, stop, stride = arg.indices(num_rows)
+        if start < 0:
+            start = start + num_rows
+        if stop < 0:
+            stop = stop + num_rows
+
+        if start > stop:
+            return self._empty_like()
+        else:
+            start = len(self) if start > num_rows else start
+            stop = len(self) if stop > num_rows else stop
+
+            result = self._from_table(
+                libcudfxx.copying.table_slice(self, [start, stop])[0]
+            )
+
+            result._copy_categories(self)
+            return result
+
+    def _normalize_columns_and_scalars_type(self, other):
+        if isinstance(other, cudf.DataFrame):
+            return [
+                other[other_col].astype(self[self_col].dtype)._column
+                for self_col, other_col in zip(self.columns, other.columns)
+            ]
+
+        elif isinstance(other, (cudf.Series, cudf.Index)):
+            return other.astype(self.dtype)._column
+
+        else:
+            if is_scalar(other):
+                if (other is not None and not np.isnan(other)) and (
+                    self.dtype.type(other) != other
+                ):
+                    raise TypeError(
+                        "Cannot safely cast non-equivalent {} to {}".format(
+                            type(other).__name__, self.dtype.name
+                        )
+                    )
+
+                return (
+                    self.dtype.type(other)
+                    if (other is not None and not np.isnan(other))
+                    else other
+                )
```

Comment:
May want to break this off into a helper function somewhere and then reuse it here and then use it in a list comprehension below